src/Type/Php/HashHmacFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,64 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Php\PhpVersion;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\Accessory\AccessoryNonEmptyStringType;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\NeverType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function count;
+use function hash_hmac_algos;
+use function in_array;
+
+class HashHmacFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function __construct(private PhpVersion $phpVersion)
+	{
+	}
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'hash_hmac';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
+	{
+		$args = $functionCall->getArgs();
+
+		if (count($args) < 3) {
+			return null;
+		}
+
+		$algo = $scope->getType($args[0]->value);
+
+		if (!$algo instanceof ConstantStringType) {
+			$types = [new AccessoryNonEmptyStringType()];
+
+			if (!$this->phpVersion->throwsValueErrorForInternalFunctions()) {
+				$types[] = new ConstantBooleanType(false);
+			}
+
+			return TypeCombinator::union(...$types);
+		}
+
+		$knownAlgos = hash_hmac_algos();
```

Comment:
this might require a `function_exists` check, in case the extension is not loaded

Reply:
Phpstan can analyse code for extensions without the extensions beeing installed. It knows the signatures/return types.

IsFunctionSupported() only tells whether this extension here wants to handle a given call.